src/main/AllocationTable.cpp

```diff
@@ -0,0 +1,334 @@
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <limits>
+#include <map>
+
+#include "rho/AllocationTable.hpp"
+#include "rho/GCNodeAllocator.hpp"
+#include "rho/AddressSanitizer.hpp"
+
+// The low pointer bits are masked out in the hash function.
+#define LOW_BITS (19)
+#define MAX_COLLISIONS (6)
+#define MAX_REBALANCE_COLLISIONS (3)
+
+// Constant indicating a deleted bucket.
+#define EMPTY_KEY (0)
+#define DELETED_KEY (UINTPTR_MAX)
+
+typedef std::uint64_t u64;
+typedef std::uint32_t u32;
+
+using std::function;
+using std::size_t;
+
+using rho::FreeListNode;
+using rho::AllocationTable;
+typedef rho::AllocationTable::HashBucket HashBucket;
+
+typedef rho::AllocatorSuperblock Superblock;
+
+AllocationTable::AllocationTable(unsigned num_bits): num_bits(num_bits) {
+  num_buckets = 1 << num_bits;
+  hash_mask = num_buckets - 1;
+  buckets = new HashBucket[num_buckets];
+  for (int i = 0; i < num_buckets; ++i) {
+    buckets[i].data = EMPTY_KEY;
+  }
+}
+
+void AllocationTable::insert(uintptr_t pointer, size_t size) {
+  while (!try_insert(buckets, pointer, size)) {
+    // Remove partially inserted entries.
+    erase(pointer, size);
+    // Too many collisions, resize the hash table.
+    unsigned new_table_bits = num_bits + 1;
+    while (!resize(new_table_bits)) {
+      new_table_bits += 1;
+    }
+  }
+}
+
+void AllocationTable::erase(uintptr_t pointer, size_t size) {
+  uintptr_t key = pointer >> LOW_BITS;
+  uintptr_t key_end = (pointer + (1 << size) + ((1 << LOW_BITS) - 1))
+      >> LOW_BITS;
+  bool first = true;
+  pointer &= ~uintptr_t{3};  // Mask out the flag bits for equality comparison.
+  do {
+    unsigned hash = pointer_hash(key << LOW_BITS);
+    for (int i = 0; i < MAX_COLLISIONS; ++i) {
+      HashBucket& bucket = buckets[hash];
+      if (bucket.DataPointer() == pointer) {
+        bucket.data = DELETED_KEY;
+        break;
+      }
+      if (bucket.data == EMPTY_KEY) {
+        // No more partial entries inserted, we are done.
+        return;
+      }
+      hash = probe_func(hash, i);
+    }
+    key += 1;
+  } while (key < key_end);
+}
+
+HashBucket* AllocationTable::search(uintptr_t candidate) {
+  unsigned hash = pointer_hash(candidate);
+  for (int i = 0; i < MAX_COLLISIONS; ++i) {
+    HashBucket* bucket = &buckets[hash];
+    if (bucket->data != EMPTY_KEY && bucket->data != DELETED_KEY) {
+      uintptr_t pointer = bucket->DataPointer();
+      if (bucket->data & 2) {
+        if (pointer <= candidate
+            && pointer + MEDIUM_SUPERBLOCK_SIZE > candidate) {
+          // Pointer is inside this superblock. Test if it is after the header.
+          if (candidate >= pointer + SUPERBLOCK_HEADER_SIZE) {
+            return bucket;
+          } else {
+            return nullptr;
+          }
+        }
+      } else if (pointer <= candidate
+            && (pointer + (1L << bucket->size)) > candidate) {
```

Comment:
The calculation of the end of the memory in the allocation should be a HashBucket / Allocation method.